feat(altda-client): punctuality check via passing l1_inclusion_block_num to da-server

samlaf opened the pull request:
Not quite sure what we'll end up calling this check: punctuality, recency, timing, etc.
But main idea is that the derivation pipeline needs a way to prevent a malicious batcher from submitting an old cert whose data is no longer available on the DA Layer from stalling the derivation pipeline. So the rollup needs to have a rule to discard these certs as invalid. See https://github.com/Layr-Labs/eigenda/pull/1533 for discussions related to this.

This change simply passes the l1_inclusion_block_number at which a commitment was included in the batcher inbox as a query param to GET requests submitted to da-server, where the da-server can encode the rule that different rollups/da-layers want to enforce. Here's the [related PR](https://github.com/Layr-Labs/eigenda-proxy/pull/378) for eigenda-proxy to support this query param.

TODO: add tests

samlaf committed:
feat(altda-client): pass l1_inclusion_block_number as query param to da server

This is used to perform punctuality check on EigenDA, but is generic and should be used by all da layers for the same purpose.

bxue-l2 commented on op-alt-da/daclient.go:
```diff
@@ -42,8 +46,10 @@ func NewDAClient(url string, verify bool, pc bool) *DAClient {
+// The l1InclusionBlock at which the commitment was included in the batcher-inbox is submitted to the DA server.
+// It is used to discard old commitments whose blobs have a risk of not being available anymore.
+func (c *DAClient) GetInput(ctx context.Context, comm CommitmentData, l1InclusionBlock eth.L1BlockRef) ([]byte, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, fmt.Sprintf("%s/get/0x%x?l1_inclusion_block_number=%d", c.url, comm.Encode(), l1InclusionBlock.Number), nil)
```
when it is ready, can we parse the error code from the request, such that it is ok for op-node to skip some bad cert according to proxy

samlaf replied:
[646c885](https://github.com/Layr-Labs/optimism/pull/45/commits/646c8857f4b73e67682e1b60ca14680b2c79d2ba)

samlaf committed:
feat(altda): drop invalid certs

Defined generic protocol where da-server will return a 418 (TEAPOT) error when a cert is invalid, along with the reason (250 bytes) in the body.

The 418 error is transformed into an internal golang InvalidCommitmentError which when received by the derivation pipeline, causes it to skip the commitment and move forward.

bxue-l2 commented on op-alt-da/daclient.go:
```diff
@@ -61,6 +73,14 @@ func (c *DAClient) GetInput(ctx context.Context, comm CommitmentData, l1Inclusio
 	if resp.StatusCode == http.StatusNotFound {
 		return nil, ErrNotFound
 	}
+	if resp.StatusCode == http.StatusTeapot {
```
It looks like ErrNotFound will trigger retry inside altda_data_source in derivation pipeline

samlaf replied:
it this a comment on my code or just a random comment?
This is why we need to use a new error code (StatusTeapot and InvalidCommitmentError) for certVerification failures, because StatusNotFound/ErrNotFound was already being used for keccak stuff.

bxue-l2 commented on op-node/rollup/derive/altda_data_source.go:
```diff
@@ -91,6 +92,10 @@ func (s *AltDADataSource) Next(ctx context.Context) (eth.Data, error) {
+	} else if errors.As(err, &invalidCommitmentError) {
+		s.log.Warn("skipping invalid commitment", "comm", s.comm, "err", err)
+		s.comm = nil
+		return s.Next(ctx) // skip the input
```
this is handy. I was thinking it should return an empty byte. But it looks like it is eliminated entirely, before going to the next stage of derivation pipeline. I.e it is entirely handled at this stage. I think hokulea should use the same strategy

bxue-l2 replied:
I think we need to be extra carefully about aligning the behavior of the go and hokulea, I think it is somewhere here https://github.com/Layr-Labs/hokulea/blob/a21c33d25b7297d6c77a5f62da386565ee0fe90c/crates/eigenda/src/eigenda_blobs.rs#L89 

I think the behavior is roughly aligned, but good to have extra eye. Also the error propagation in hokulea is pretty bad

samlaf committed:
chore(daclient): use uint64 for blocknum directly instead of L1BlockRef struct

The struct was confusing to use in tests because it wasnt sure only the .Number field of it was used, so made implementers unsure whether they needed to populate the whole struct. Since we only used the .Number field, I opted to just take a uint64 directly as argument to GetInput.

litt3 commented on op-alt-da/daclient.go:
```diff
@@ -55,6 +73,14 @@ func (c *DAClient) GetInput(ctx context.Context, comm CommitmentData) ([]byte, e
 	}
+	if resp.StatusCode == http.StatusTeapot {
+		defer resp.Body.Close()
+		var invalidCommitmentReason [250]byte
```
What's the significance of 250?

samlaf replied:
Arbitrary number that I guessed should be enough. I will add a comment. Think its always better to have bounds though, even if arbitrary.

samlaf replied:
[9251a34](https://github.com/Layr-Labs/optimism/pull/45/commits/9251a34831a789525f5e4c726ed63e5a81ee19d6)

litt3 commented on op-alt-da/daclient.go:
```diff
@@ -55,6 +73,14 @@ func (c *DAClient) GetInput(ctx context.Context, comm CommitmentData) ([]byte, e
+		// We discard the error as it only contains the reason for invalidity.
+		// We might end up with a partial reason, but the commitment should still be skipped.
+		_, _ = io.ReadFull(resp.Body, invalidCommitmentReason[:])
+		return nil, InvalidCommitmentError{Reason: string(invalidCommitmentReason[:])}
```
Would it make sense to capture the exact number of bytes read? Something like this:

```
		buf := make([]byte, 250)
		n, _ := io.ReadFull(resp.Body, buf)
		return nil, InvalidCommitmentError{Reason: string(buf[:n])}
```

samlaf replied:
Yes good catch! Changed code to use an `http.MaxBytesReader` instead: [9251a34](https://github.com/Layr-Labs/optimism/pull/45/commits/9251a34831a789525f5e4c726ed63e5a81ee19d6)

bxue-l2 reviewed (approved):
Btw, l1blockref contain timestamp. But uint64 looks good for now

samlaf committed:
style(daclient): use MaxBytesReader

Use MaxBytesReader and document why we restrict error body to 1000 characters (upped from 250 previously): to prevent DDoS
